src/memory_region.rs

```diff
@@ -33,52 +33,93 @@ use std::{
 */
 
 /// Callback executed before generate_access_violation()
-pub type MemoryCowCallback = Box<dyn Fn(u32) -> Result<u64, ()>>;
+///
+/// Safety: Reentrancy is forbidden.
+pub type AccessViolationHandler = Box<dyn Fn(&mut MemoryRegion, u64, AccessType, u64, u64)>;
 /// Fail always
 #[allow(clippy::result_unit_err)]
-pub fn default_memory_cow_callback(_cow_callback_payload: u32) -> Result<u64, ()> {
-    Err(())
+pub fn default_access_violation_handler(
+    _region: &mut MemoryRegion,
+    _region_max_len: u64,
+    _access_type: AccessType,
+    _vm_addr: u64,
+    _len: u64,
+) {
+}
+macro_rules! access_violation_guard {
+    ($self_ty:ty, $self:expr, $access_type:expr, $vm_addr:expr, $len:expr) => {{
+        if let Some((index, region)) = $self.find_region($vm_addr) {
+            if let Some(host_addr) = region.vm_to_host($access_type, $vm_addr, $len) {
+                return ProgramResult::Ok(host_addr);
+            }
+            {
```

Comment:
This does not need to be wrapped in brackets.

Reply:
That is for the scope guard of the `RefCell` borrow.